### PR TITLE
Pedro ac2

### PR DESCRIPTION
**Pull request by prchall123**
- Added sun pointing and detumbling controllers.
- Fixed attitude kynematics bug
- Added plots to post-processing
- Added sun position and magnetic field to state vector

### CODE REVIEW

File: FSW/controllers/controller.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            ref_torque = np.zeros((3,))
+        elif pointing_mode == "3D-stabilized":
+            q_ref = rotmat2quat(Rot_mat)
+            ref_ctrl_states[:4] = q_ref
+            ref_torque = np.zeros((3,))
+        else:
+            raise ValueError(f"Unrecognized pointing mode: {pointing_mode}")
+            
+        return ref_torque, ref_ctrl_states
+
```
Comment (Herpderk) [inline FSW/controllers/controller.py:107]: Rather than checking for the control alg ID, I think it makes sense to initialize a class attribute that points to the control alg callable. So something like
``` python
# on object initialization, point to the callable without passing args
if alg_id == "Bcross":
    self.my_control_alg  = self.lyapsunpointctr.get_dipole_moment_command
```

File: FSW/controllers/controller.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            mtb_torque = self.lyapsunpointctr.get_dipole_moment_command(est_ctrl_states, Idx) 
+            return mtb_torque, []
+        elif self.controller_algo == "BaseSP":
+            mtb_torque = self.basesunpointctr.get_dipole_moment_command(est_ctrl_states, Idx) 
+            return mtb_torque, []
+        elif self.controller_algo == "BaseNP":
+            mtb_torque, rw_torque = self.nadirpointctr.get_dipole_moment_command_and_rw_torque(est_ctrl_states, Idx) 
+            return mtb_torque, rw_torque
+        else: 
+            raise ValueError(f"Unrecognized controller algorithm: {self.controller_algo}")
```
Comment (Herpderk) [inline FSW/controllers/controller.py:123]: Rather than checking for the control algorithm identifier everytime we want torque, I think it makes sense to initialize a class attribute that points to the control algorithm callable. So something like
``` python
# on object initialization, point to the callable without passing args
if alg_id == "Bcross":
    self.my_control_alg  = self.lyapsunpointctr.get_dipole_moment_command
```

Comment (Herpderk) [reply FSW/controllers/controller.py:123]:
``` python
# everytime we want the control input
self.my_control_alg(my_args)
```

File: FSW/controllers/controller.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        
+     
+        self.allocation_mat = np.zeros((Idx["NU"],3))
+        self.allocation_mat[Idx["U"]["MTB_TORQUE"],:] = np.linalg.pinv(self.G_mtb_b.T)
+                
+        # np.linalg.pinv(B_mat)
+        actuator_cmd = np.zeros((Idx["NU"],1))
+        actuator_cmd[Idx["U"]["MTB_TORQUE"]] = np.linalg.pinv(self.G_mtb_b.T) @ mtb_torque_cmd
+        if rw_torque_cmd:
+            actuator_cmd[Idx["U"]["RW_TORQUE"]] = rw_torque_cmd # only one, bypass allocation matrix
```
Comment (Herpderk) [inline FSW/controllers/controller.py:139]: Just wondering, is this dictionary interface the standard way of getting vectors across our whole sim? I like some things about it, but I think we can do better in making it more robust

File: FSW/controllers/controller.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                                                    self.G_mtb_b,
+                                                    config["bcrossgain"],
+                                                    np.array(config["mtb_att_feedback_gains"]),
+                                                    np.array(config["rw_att_feedback_gains"]),
+                                                    np.array(config["rw_vel_gain"]),
+                                                    np.array(config["tgt_ss_ang_vel"]),
+                                                    np.array(config["nom_rw_ang_vel"]),
+                                                    np.array(config["nadir_cam_dir"]),
+                                                    Magnetorquers,
+                                                    ReactionWheels)
```
Comment (Herpderk) [inline FSW/controllers/controller.py:44]: 2 issues:
- I think we should only initialize one type of controller per generic controller instance. Ideally, every specialized controller inherits abstract methods from the base controller and adheres to some standard interfacing method dictated by the parent class.
- The constructor arguments are just not very readable. It would be better if there was standard object type being passed in (e.g. all numpy arrays), or if we pack all related data together neatly into a single class and pass that in (as is the case with the `Magnetorquers`).

Comment (Herpderk) [reply FSW/controllers/controller.py:44]: But these are just changes for long-term maintainability. If it's too much work, let's make a new branch for that after merging these changes

File: FSW/controllers/controller.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        if rw_torque_cmd:
+            actuator_cmd[Idx["U"]["RW_TORQUE"]] = rw_torque_cmd # only one, bypass allocation matrix
+        # np.linalg.pinv(self.G_rw_b.T) @ rw_torque_cmd
+        # Normalize columns of the allocation matrix
+        # col_norms = np.linalg.norm(allocation_mat, axis=0)
+        # allocation_mat = allocation_mat / col_norms
+        
+        return actuator_cmd
+    
+    def run(self, date, est_world_states, Idx):
```
Comment (Herpderk) [inline FSW/controllers/controller.py:147]: rename to something more descriptive, like `get_voltages`

File: FSW/controllers/controller.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        
+        self.est_world_states = est_world_states
+        # define slew profile
+        ref_torque, ref_ctrl_states = self.define_att_profile(date, self.pointing_mode, self.pointing_target, 
+                                                              self.tgt_ang_vel, est_world_states)
```
Comment (Herpderk) [inline FSW/controllers/controller.py:152]: do we really need to define our attitude profile every time we ask for the control input? Similar issue with checking for the control algorithm type at every call of `run`. IMO the objectives and algorithms of the controller objects should be given clear distinctions on startup, and inherit from a common set of interfacing methods.

File: FSW/controllers/current_distributor.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        k = np.array([magnetorquers[0].get_dipole_moment_over_current()])
+
+        for i in range(1, len(magnetorquers)):
+            U = np.column_stack((U, magnetorquers[i].G_mtb_b))
+            k = np.hstack((
+                k, magnetorquers[i].get_dipole_moment_over_current()
+            ))
+
+        K = np.diag(k)
+        return np.linalg.pinv(U @ K)
```
Comment (Herpderk) [inline FSW/controllers/current_distributor.py:32]: We need to change everything to voltage because that's how we're interfacing the coils IRL

**Comment by Herpderk** [comment]
In general, there should be more type hinting and maybe even a README describing the methods that the sim directly interfaces with. I also advocate for a class attribute naming convention that distinguishes "public" from "private" members and properties--e.g. `object.public_member` v.s. `object._private_member`.

**Review by Herpderk — APPROVED** [review]
The FSW is pretty messy but let's change that for the next PR. If it works it works.

File: FSW/controllers/controller.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        elif pointing_mode == "3D-stabilized":
+            q_ref = rotmat2quat(Rot_mat)
+            ref_ctrl_states[:4] = q_ref
+            ref_torque = np.zeros((3,))
+        else:
+            raise ValueError(f"Unrecognized pointing mode: {pointing_mode}")
+            
+        return ref_torque, ref_ctrl_states
+    
+    def get_torque(self, date, ref_torque, ref_ctrl_states, est_ctrl_states, Idx):
```
Comment (kmkar31) [inline FSW/controllers/controller.py:108]: Why do we need all these arguments? We'd want an argument that provides the current measurement / state estimate vector and we compute the MTB voltages. Reference states etc. need to be class attributes.

Secondly, we need docstrings for these methods esp. if you're using variables called 'idx'

File: FSW/controllers/controller.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        # if torque = B @ actuator_cmd
+        # actuator_cmd = Binv @ torque_cmd
+        
+     
+        self.allocation_mat = np.zeros((Idx["NU"],3))
+        self.allocation_mat[Idx["U"]["MTB_TORQUE"],:] = np.linalg.pinv(self.G_mtb_b.T)
+                
+        # np.linalg.pinv(B_mat)
+        actuator_cmd = np.zeros((Idx["NU"],1))
+        actuator_cmd[Idx["U"]["MTB_TORQUE"]] = np.linalg.pinv(self.G_mtb_b.T) @ mtb_torque_cmd
```
Comment (kmkar31) [inline FSW/controllers/controller.py:137]: I don't know how badly constrained we are with memory, but we should probably just use lists which contain moments in a pre-defined order

File: FSW/controllers/detumbling_controller.py [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+        self.J = inertia_tensor
+        self.dt = time_step
+        self.alpha = alpha
+        self.beta = beta
+        self.lbm = dipole_moment_lower_bound
+        self.ubm = dipole_moment_upper_bound
+
+    def get_dipole_moment_command(
+        self,
+        magnetic_field_in_eci_frame: np.ndarray,
```
Comment (kmkar31) [inline FSW/controllers/detumbling_controller.py:40]: FSW does not get to dictate which frame the arguments are provided in. It only receives sensor data and we cannot add a pre-processing step. So we should assume all quantites are in the body frame (except some sensors like star-tracker that operate in ECI)

File: FSW/controllers/detumbling_controller.py [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+            Bhat_pseudo_inv = V.T @ S_inv @ U.T
+
+            err_vec = np.hstack((-self.I_min_direction-sun_vector, self.ref_angular_velocity - angular_velocity))
+            u = Bhat_pseudo_inv @ self.k @ err_vec
+
+        u = np.clip(a=u, a_min=self.lbm, a_max=self.ubm)
+        """
+        print(f"Spin-stabilizing: h = {h}, Norm of angular momentum h_norm = {h_norm}")
+        print("torque command: ", crossproduct(u) @ magnetic_field)
+        angle_sun_h = np.arccos(np.clip(np.dot(sun_vector, h / h_norm), -1.0, 1.0))
```
Comment (kmkar31) [inline FSW/controllers/detumbling_controller.py:237]: Delete Debug print statements

File: FSW/controllers/detumbling_controller.py [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+            """
+            U, S, V = np.linalg.svd(Bhat)
+            idx = np.where(S > 1e-6 * np.max(S))[0]
+            S_inv = np.diag([1/s if s > 1e-10 else 0 for s in S])
+            Bhat_pseudo_inv = V.T @ S_inv @ U.T"""
+            Bhat_pseudo_inv = Bhat.T / (magnetic_field_norm ** 2)
+            # 1. mtb nutation control
+            err_vec = np.hstack((-self.G_rw_b.flatten()-orbit_vector, angular_velocity))
+            u_mtb = Bhat_pseudo_inv @ self.k_mtb_att @ err_vec
+            u_mtb = np.clip(a=u_mtb, a_min=self.lbmtb, a_max=self.ubmtb)
```
Comment (kmkar31) [inline FSW/controllers/detumbling_controller.py:375]: We really need to see how much time this takes. So many array ops will crash PyCubed

File: actuators/Magnetorquer.cpp [hunk 13]
```diff
@@ -2,53 +2,64 @@
...
+#include <utility>
+#include <iostream>
 
-
-Magnetorquer::Magnetorquer(int N_MTBs, double maxVolt, double coilsPerLayer, double layers, double traceThickness,
-                           double traceWidth, double gapWidth, double maxPower, double maxCurrentRating,
-                           MatrixXd mtb_orientation) 
+Magnetorquer::Magnetorquer(int N_MTBs, VectorXd maxVolt, VectorXd coilsPerLayer, VectorXd layers, VectorXd traceThickness,
+                           VectorXd pcbSideMax, VectorXd traceWidth, VectorXd gapWidth, VectorXd maxPower, VectorXd maxCurrentRating,
+                           MatrixXd mtb_orientation) : num_MTBs(N_MTBs), max_voltage(std::move(maxVolt)), N(std::move(coilsPerLayer)), pcb_layers(std::move(layers)),
```
Comment (kmkar31) [inline actuators/Magnetorquer.cpp:10]: I get that we want to account for variations in MTBs in Monte-Carlo, but do we really need to perturb all these quantities? Won't simply perturbing Coil resistance do? I.e., 
$$R_{mtb}^{i=1:N_{mtb}} =R_{nominal}+\Delta R_{i}^{i=1:N_{mtb}}$$

where $R_{nominal}$ is a double and the other two are vectors

File: actuators/magnetorquer.py
Comment (kmkar31) [inline actuators/magnetorquer.py:1]: Needs to be deleted. Obsolete file

File: actuators/reaction_wheels.py
Comment (kmkar31) [inline actuators/reaction_wheels.py:1]: Needs to be deleted. Obsolete File

File: config.yaml
Comment (kmkar31) [inline config.yaml:1]: Needs to be deleted. Obsolete File

File: montecarlo/configs/params.yaml [hunk 14]
```diff
@@ -15,39 +18,54 @@ I_rw : 6.5e-7              # kg.m^2 inertia of each reaction wheel
...
+AOP : 63 # [deg]
 true_anomaly : 0 # [deg]
 initial_attitude : [1,0,0,0] 
-initial_angular_rate : [0.3,0,0] # [rad/s]
+initial_angular_rate : [0.2,0.3,0.2] # [rad/s]
+
+algorithm: "BaseNP" # "BaseSP" # "Lyapunov" # "Bcross" # # "BCROSS" or "PID" or "BaseSP"
+bcrossgain: 0.0028
 
+mtb_att_feedback_gains: [[0.7071, 0.0, 0.0, 0.0028, 0.0, 0.0],
```
Comment (kmkar31) [inline montecarlo/configs/params.yaml:62]: Controller gains should be written into the controller.py file. This yaml file will not be a part of FSW

File: requirements.txt [hunk 15]
```diff
@@ -1,4 +1,6 @@
 numpy
 matplotlib
 basemap
-psutil
\ No newline at end of file
+psutil
+brahe
```
Comment (kmkar31) [inline requirements.txt:5]: Why are we reusing brahe?

File: sensors/sensors.py
Comment (kmkar31) [inline sensors/sensors.py:1]: I don't get why unrelated files are being renamed. This seems like branches being jumbled

File: sim.py [hunk 16]
```diff
@@ -6,69 +6,117 @@
...
+                "q_w", "q_x", "q_y", "q_z", "omega_x [rad/s]", "omega_y [rad/s]", "omega_z [rad/s]",
+                "rSun_x ECI [m]","rSun_y ECI [m]","rSun_z ECI [m]","xMag ECI [T]","yMag ECI [T]","zMag ECI [T]"] + ["omega_RW_" + str(i) + " [rad/s]" for i in range(num_RWs)]
     #measurement_labels = state_labels # TODO : Fix based on partial state measurement
     state_estimate_labels = ["r_hat_x ECI [m]", "r_hat_y ECI [m]", "r_hat_z ECI [m]", "v_hat_x ECI [m/s]", "v_hat_y ECI [m/s]", "v_hat_z ECI [m/s]",
-                "q_hat_w", "q_hat_x", "q_hat_y", "q_hat_z", "omega_hat_x [rad/s]", "omega_hat_y [rad/s]", "omega_hat_z [rad/s]"] + ["omega_hat_RW_" + str(i) + " [rad/s]" for i in range(num_RWs)]
-    input_labels = ["V_MTB_" + str(i) + " [V]" for i in range(num_MTBs)] + ["V_RW_" + str(i) + " [V]" for i in range(num_RWs)]
-
+                "q_hat_w", "q_hat_x", "q_hat_y", "q_hat_z", "omega_hat_x [rad/s]", "omega_hat_y [rad/s]", "omega_hat_z [rad/s]",
+                "rSun_x ECI [m]","rSun_y ECI [m]","rSun_z ECI [m]","xMag ECI [T]","yMag ECI [T]","zMag ECI [T]"] + ["omega_hat_RW_" + str(i) + " [rad/s]" for i in range(num_RWs)]
```
Comment (kmkar31) [inline sim.py:51]: Having Sun pos and Mag field as a part of the true state vector makes sense. Why is it also a part of the state estimate? I don't think FSW will be filtering the sun position and magnetometer readings

File: sim.py [hunk 17]
```diff
@@ -6,69 +6,117 @@
...
 def run(log_directory, config_path):
     logr = logger.MultiFileLogger(log_directory)
 
-    params = SimParams(config_path)
+    params = SimParams(config_path) # cpp params
+    with open(config_path, "r") as f:
+        pyparams = yaml.safe_load(f)
 
     initial_state = np.array(params.initial_state) # Get initial state
+    newstate = np.array(rk4(initial_state, np.zeros((params.num_MTBs + params.num_RWs, 1)), params, -params.dt, params.dt))
```
Comment (kmkar31) [inline sim.py:39]: This needs to go into C++ initializeSatellite() function

File: simulation_manager/run_job.py [hunk 18]
```diff
@@ -14,7 +14,7 @@
 
 # ============================================= JOB CONFIGURATION CONSTANTS ===========================================
 NUM_TRIALS = 1
-TIMEOUT = 300  # [s] kill a trial that takes longer than this amount
+TIMEOUT = 900  # [s] kill a trial that takes longer than this amount
```
Comment (kmkar31) [inline simulation_manager/run_job.py:17]: Is controller computation taking that much longer than dynamics? Python is slow, but this is a huge increase

File: simulation_utils/ParameterParser.cpp [hunk 19]
```diff
@@ -77,14 +81,21 @@ void Simulation_Parameters::getParamsFromFileAndSample(std::string filename) {
...
     Vector6 KOE {semimajor_axis, eccentricity, inclination, RAAN, AOP, true_anomaly};
     Vector6 CartesianState = KOE2ECI(KOE, epoch);
     State(Eigen::seqN(0,6)) = CartesianState;
     State(Eigen::seqN(6,4)) = initial_attitude;
     State(Eigen::seqN(10,3)) = initial_angular_rate;
-    State(Eigen::seqN(13, num_RWs)).setZero();
+    // Sun Position [m]
+    State(Eigen::seqN(13, 3)) = Eigen::VectorXd::Zero(3); //sun_position_eci(epoch);
+    // Magnetic Field [T]
+    // Vector3 Bfield = MagneticField(CartesianState(Eigen::seqN(0, 3)), epoch);
```
Comment (kmkar31) [inline simulation_utils/ParameterParser.cpp:94]: Why is this commented out? This is the best way to load it into the state

File: world/math/quaternions.py
Comment (kmkar31) [inline world/math/quaternions.py:1]: Obsolete. Delete

File: world/math/transforms.py
Comment (kmkar31) [inline world/math/transforms.py:1]: This file must be deleted

File: world/physics/attitude.py
Comment (kmkar31) [inline world/physics/attitude.py:1]: Needs to be deleted

**Review by kmkar31 — CHANGES_REQUESTED** [review]
1. There are a lot of changes to be made to bring this into a merge-able version
2. Even without architecture / comment -related changes, I'm afraid there's been some branch mixing and we'd spend more time fixing these issues than adding capabilities

It'd be great if you could pull main, copy paste your additional changes and then force push to this branch.

File: FSW/controllers/controller.py (hunk 3 above)
Comment (prchall123) [reply FSW/controllers/controller.py:139]: we can adopt another structure for it if you think there's a better one. This code won't be FSW, it's only there to make it easier to read than to have to backtrack everytime to understand what variables are being taken from the state vectors

File: simulation_utils/ParameterParser.cpp (hunk 19 above)
Comment (prchall123) [reply simulation_utils/ParameterParser.cpp:94]: wasnt working. Need to recheck

File: simulation_manager/run_job.py (hunk 18 above)
Comment (prchall123) [reply simulation_manager/run_job.py:17]: no, this was because I was running really long sims.

File: FSW/controllers/detumbling_controller.py (hunk 12 above)
Comment (prchall123) [reply FSW/controllers/detumbling_controller.py:375]: this won't make it to FSW. This is our Model-In-the-loop code that only serves to prototype algorithms and to min trouble with sw dev. Only concern here is code being readable and easy to prototype new algos

File: FSW/controllers/detumbling_controller.py (hunk 10 above)
Comment (prchall123) [reply FSW/controllers/detumbling_controller.py:40]: this is not supposed to be the actual fsw, only modelling what it does. but yeah that there is not correct

File: FSW/controllers/controller.py (hunk 9 above)
Comment (prchall123) [reply FSW/controllers/controller.py:137]: we can precompute this in the class and load it every time. This isnt the actual flight code, so for these sims in particular its not an issue. For SIL it will be and we will have to do what you describe

File: FSW/controllers/controller.py (hunk 8 above)
Comment (prchall123) [reply FSW/controllers/controller.py:108]: some of these may no longer be necessary. I'll revisit this

File: FSW/controllers/current_distributor.py (hunk 7 above)
Comment (prchall123) [reply FSW/controllers/current_distributor.py:32]: I wasnt using the current distributor function, but the controller function was doing that.

File: FSW/controllers/controller.py (hunk 6 above)
Comment (prchall123) [reply FSW/controllers/controller.py:152]: the function isnt used anymore, we can ditch it. I thought we'd need a guidance algorithm, but we can get away with having everything in the controller.

File: actuators/Magnetorquer.cpp (hunk 13 above)
Comment (prchall123) [reply actuators/Magnetorquer.cpp:10]: yes, perturbing R is enough. We can change interfaces here

File: montecarlo/configs/params.yaml (hunk 14 above)
Comment (prchall123) [reply montecarlo/configs/params.yaml:62]: controller.py wont be a part of fsw either

File: requirements.txt (hunk 15 above)
Comment (prchall123) [reply requirements.txt:5]: possibly no longer needed, need to check

File: sensors/sensors.py
Comment (prchall123) [reply sensors/sensors.py:1]: I don't get what you mean here

File: sim.py (hunk 16 above)
Comment (prchall123) [reply sim.py:51]: we need a measurement vector, but ad hasnt been implemented. we can put them there.

File: FSW/controllers/controller.py (hunk 8 above)
Comment (prchall123) [reply FSW/controllers/controller.py:108]: also I have pulled main into this branch

File: actuators/Magnetorquer.cpp (hunk 13 above)
Comment (kmkar31) [reply actuators/Magnetorquer.cpp:10]: Are we making this change in this PR?

File: sim.py (hunk 16 above)
Comment (kmkar31) [reply sim.py:51]: I will add a PR that adds sensor models after this one

File: FSW/controllers/controller.py (hunk 2 above)
Comment (prchall123) [reply FSW/controllers/controller.py]: changed. If you can lmk if the latest version seems better

File: FSW/controllers/controller.py (hunk 4 above)
Comment (prchall123) [reply FSW/controllers/controller.py]: Ive created a super class controller algorithm. We can edit further in the future

File: FSW/controllers/controller.py (hunk 5 above)
Comment (prchall123) [reply FSW/controllers/controller.py:78]: I wouldnt change the run function's name. maybe a subfunction

File: FSW/controllers/controller.py (hunk 6 above)
Comment (prchall123) [reply FSW/controllers/controller.py]: function has been ditched

File: FSW/controllers/detumbling_controller.py (hunk 10 above)
Comment (prchall123) [reply FSW/controllers/detumbling_controller.py]: this is part of the nonmonotonic detumbling controller which we won't be using

File: FSW/controllers/detumbling_controller.py (hunk 11 above)
Comment (prchall123) [reply FSW/controllers/detumbling_controller.py]: done

File: world/math/quaternions.py
Comment (prchall123) [reply world/math/quaternions.py:1]: I need quaternion math in python

File: world/physics/attitude.py
Comment (prchall123) [reply world/physics/attitude.py:1]: done

File: simulation_utils/ParameterParser.cpp (hunk 19 above)
Comment (prchall123) [reply simulation_utils/ParameterParser.cpp:95]: I dont know how to fix this. Theres some spice bug I dont know how to solve. I would ask you to help fix it whenever you find the time

File: actuators/Magnetorquer.cpp (hunk 13 above)
Comment (prchall123) [reply actuators/Magnetorquer.cpp:10]: hmm I dont think it should be blocking us at this point. We can change this later, lets worry about having a single integrated and validated sim first

File: sensors/sensors.py
Comment (prchall123) [reply sensors/sensors.py:1]: Ive pulled and updated the branch with main.

File: sim.py (hunk 17 above)
Comment (prchall123) [reply sim.py:39]: cant make it work. Ill need help on this, Ive spent too much time and its not critical

**Comment by prchall123** [comment]
All comments have been addressed to the extent I'm able to deal with them. @kmkar31 I'd ask you to review if youre happy with the current version.

File: world/math/quaternions.py
Comment (kmkar31) [inline world/math/quaternions.py:1]: This should go into utils_and_transforms.cpp, but can be done in a later PR

**Review by kmkar31 — COMMENTED** [review]
Just a few files to be deleted. No other changes

File: actuators/magnetorquer.py
[the remark this answers is not in the document]
Comment (prchall123) [reply actuators/magnetorquer.py:1]: used by the controller, but we can probably get rid of it. I would maybe suggest having those changes done when we prep the sim for MC campaigns

File: actuators/reaction_wheels.py
[the remark this answers is not in the document]
Comment (prchall123) [reply actuators/reaction_wheels.py:1]: yes, same for the previous one

File: config.yaml
[the remark this answers is not in the document]
Comment (prchall123) [reply config.yaml:1]: yes youre right, ill delete it

File: world/math/quaternions.py
Comment (prchall123) [reply world/math/quaternions.py:1]: we would want a python utils equivalent though

**Review by kmkar31 — APPROVED** [review]
Approved
